lcls_tools/common/data_analysis/bmad_modeling.py

```diff
@@ -0,0 +1,573 @@
+                    cmd_words = cmd.split()
+                    cmd_words[-1] = str(pvdata[pv])
+                    new_lines.append(" ".join(cmd_words))
+                else:
+                    new_lines.append(cmd)
+            lines_rf = new_lines
+        return lines_rf + lines_quads
+
+    def use_klys_when_beam_off(self, tao_cmds, pvdata, beam_code="1"):
+        """Modifies tao_cmds, in_use set to 1 if station active on
+        beam_code"""
+        # use = {}
+        new_cmd = []
+        for cmd in tao_cmds:
+            if 'in_use' in cmd:
+                ele = cmd.split()[2]
+                [sector, station] = ele[1:].split("_")
+                pv = f"KLYS:LI{sector}:{station}1:BEAMCODE{beam_code}_STAT"
+                # use[ele] = pvdata[pv]
+                cmd_words = cmd.split()
+                cmd_words[-1] = pvdata[pv]
+                new_cmd.append(" ".join(cmd_words))
+            else:
+                new_cmd.append(cmd)
+        return new_cmd
+
+    def get_machine_values(self, pv_list):
+        """Returns pvdata, a dictionary with keys containing
+        the PV name and values from Actual, Desired or Archive"""
+        if self.data_source in ["DES", "ACT"]:
+            pvdata = self.get_live(pv_list)
+        elif "ARCHIVE" in self.data_source:
+            pvdata = lcls_archiver_restore(pv_list, self.date_time)
+        return pvdata
+
+    def tc(self, tao, cmd):
+        """Prints result of tao.cmd() one line at a time"""
+        result = tao.cmd(cmd)
+        [print(line) for line in result]
+
+    def update_energy_gain(self, tao, pvdata, region):
+        """Updates energy gain profile based on bending magnets,
+        calculates a fudge and
+        modifies model's cavity amplitudes and phases"""
+
+        if "sc" in self.beam_path:
+            cavities = tao.lat_list(
+                f"LCAV::BEG{region}:END{region}", "ele.name")
+            for indx, cav in enumerate(cavities):
+                cavities[indx] = cav.split("#")[0]
+            cavities = list(set(cavities))
+            devices = [tao.ele_head(element)["alias"] for element in cavities]
+            if self.data_source == "ACT":
+                attr = "ACTMEAN"
+            elif self.data_source in ["DES", "ARCHIVE"]:
+                attr = "DES"
+            ampl = [pvdata[dev + ":A" + attr] for dev in devices]
+            phas = [pvdata[dev + ":P" + attr] for dev in devices]
+            amplNp = np.array(ampl)
+            amplNp = np.where(amplNp < 0.5, 0, amplNp)
+            phasNp = np.array(np.deg2rad(phas))
+            gainMeasured = amplNp * np.cos(phasNp)
+            if region == "L2B":
+                expected_gain = 1000 * (
+                    pvdata["REFS:BC2B:500:EDES"] - pvdata["REFS:BC1B:400:EDES"]
+                )
+            elif region == "L3B":
+                expected_gain = 1000 * (
+                    pvdata["REFS:DMPS:400:EDES"] - pvdata["REFS:BC2B:500:EDES"]
+                )
+
```

Comment:
Can we get some comments on what is happening here? w.r.t. fudge and the calcs.

Reply:
I changed the name of the variables to make it more readable and added a comment line that defines the LEM fudge factor
            # LEM fudge factor = 1 + dF
            dF = (expected_energy_gain - sum(measured_energy_gain)) / sum(ampl)